miprometheus/problems/question_context_to_class/babiqa_dataset_single_question.py

```diff
@@ -0,0 +1,514 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, e:wqither express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""bAbiQA.py: contains code for loading the babi dataset (based on the parsing used in torchtext)"""
+__author__= "Vincent Albouy, Ryan.L McAvoy"
+
+import torch
+from miprometheus.utils.problems_utils.language import Language
+import torch.utils.data
+from tqdm import tqdm
+import os
+from miprometheus.utils.app_state import AppState
+from miprometheus.problems.seq_to_seq.seq_to_seq_problem import SeqToSeqProblem
+from miprometheus.utils.loss.masked_cross_entropy_loss import MaskedCrossEntropyLoss
+
+
+
+class BABI(SeqToSeqProblem):
+    """
+    Problem Class for loading bAbi QA data set using Torchtext
+    
+    Inherits from SeqToSeqProblem
+    
+    """
+
+    def __init__(self, params):
+        """
+   
+        Initializes BABI QA problem, calls base class initialization, sets properties using the provided parameters.
+
+        :param params: Dictionary of parameters (read from configuration file).
+        
+        """
+        super(BABI).__init__()
+
+        self.directory = './'
+
+        # boolean: is it training phase?
+        self.data_type = params['data_type']
+
+        self.use_batches = params['batch_size']
+
+        # task number to train on
+        self.tasks = params['tasks']
+
+        self.loss_function = MaskedCrossEntropyLoss()
+
+        self.tenK = params['ten_thousand_examples']
+
+        self.one_hot_embedding = params['one_hot_embedding']
+
+        self.batch_size = params['batch_size']
+
+        self.memory_size = params['truncation_length']
+
+        self.embedding_type = params['embedding_type']
+
+        self.embedding_size = 38
+
+        self.init_token = '<sos>'
+
+        self.pad_token = '<pad>'
+
+        self.eos_token = '<eos>'
+
+        self.use_mask = False
+
+        self.urls = ['http://www.thespermwhale.com/jaseweston/babi/tasks_1-20_v1-2.tar.gz']
+
+        self.name = 'BABIDataset'
+
+        self.dirname = ''
+
+        self.data = self.load_data( tasks=self.tasks, tenK=self.tenK, add_punctuation=True , data_type = self.data_type)
+
+        #create an object language from Language class - This object will be used to create the words embeddings
+        self.language = Language('lang')
+
+        self.default_values = {'input_item_size': self.embedding_size , 'output_item_size':self.embedding_size}
+
+        self.data_definitions = {'sequences': {'size': [-1, -1, self.memory_size], 'type': [torch.Tensor]},
+                                 'targets': {'size': [-1], 'type': [torch.Tensor]},
+                                 'current_question': {'size': [-1, 1], 'type': [list, str]},
+                                 'masks': {'size': [-1], 'type': [torch.Tensor]},
+                                 }
```

Comment:
Aside of that, we got twenty tasks in bAbI. I guess it would be good to store which task qiven sample belongs to.

Similarly as in the CLEVR case, we will need that to analyze how good we are e.g. in stories with two supporting facts. Please notice that people when reporting results on bAbI are providing tables with separate accuracies on all 20 tasks and then one accuracy when trained jointly on all tasks...

Reply:
Yes ok